src/machines/java/com/enderio/machines/common/blockentity/base/MachineBlockEntity.java

```diff
@@ -411,28 +417,45 @@ public void saveAdditional(CompoundTag pTag) {
         super.saveAdditional(pTag);
 
         // Save io config.
-        pTag.put("io_config", getIOConfig().serializeNBT());
+        IIOConfig ioConfig = getIOConfig();
+        if (!ioConfig.isDefault()) {
+            pTag.put(MachineNbtTags.IO_CONFIG_NBT_KEY, getIOConfig().serializeNBT());
+        }
 
         if (supportsRedstoneControl()) {
-            pTag.putInt("redstone", redstoneControl.ordinal());
+            RedstoneControl redstoneControl = getRedstoneControl();
+            if (redstoneControl != RedstoneControl.ALWAYS_ACTIVE) {
+                pTag.putInt(MachineNbtTags.REDSTONE_CONTROL_NBT_KEY, redstoneControl.ordinal());
+            }
         }
 
         if (inventory != null) {
-            pTag.put("inventory", inventory.serializeNBT());
+            boolean hasContents = false;
```

Comment:
I don't think its necessary to do this check, for the very little storage it saves it'll add a bunch of time onto the saving process

Reply:
The reason for this check is that when the BlockEntity is converted over into an Item after being broken, it retains the extra information in the NBT data. So, for example, when a Fluid Tank is spawned/crafted into the inventory, its item NBT data is {id: "enderio:fluid_tank", Count: 1b}. When it is placed into the world, then immediately broken again, its item NBT data (without my changes) becomes: {id: "enderio:fluid_tank", Count: 1b, tag: {BlockEntityTag: {x: ..., y: ..., z: ..., io_config: ..., redstone: ..., inventory: ..., id: "enderio:fluid_tank"}}}. This has the annoying (to me at least), although admittedly not game-breaking, issue of the fact that that Fluid Tank item will not stack with another empty Fluid Tank, despite no logical reason not to.

This was simply the first way I came across of (partly) solving the issue. So I could move the functionality across into the NBT copying? But would probably need a specific provider for the Machine blocks, in the event that other things need to use that functionality.

I'll have to defer to your better judgement about the about the time in the saving process, but I could try to simplify the checks to just simple boolean checks by updating whether each item is empty/default when they are changed? Also isn't it better to potentially not save & transmit data, even at the slight increased complexity/time cost?